Contracts/guild factory

Kenny-Gin1 opened the pull request:
- Created a smart contract called `GuildRegistry` that should handle adding and removing guilds and some additional helper methods.
- Added tests (checked for 100% coverage
<img width="655" alt="Screenshot 2022-03-15 at 18 41 35" src="https://user-images.githubusercontent.com/59825171/158446911-37852d29-c155-4441-bf3f-d724e67ce14e.png">

carlbarrdahl commented on contracts/erc20guild/utils/GuildRegistry.sol:
```diff
@@ -0,0 +1,48 @@
+        uint256 guildIndexToDelete = guildStructs[guildAddress].index;
+        address guildAddressToMove = _guilds[_guilds.length - 1];
+        _guilds[guildIndexToDelete] = guildAddressToMove;
+        guildStructs[guildAddressToMove].index = guildIndexToDelete;
```
Should the mapping `guildStructs[guildAddress]` be deleted here to free up storage?

Kenny-Gin1 replied:
Hmm, I am afraid I don't understand your suggestion. Could you write out your proposal? :)

carlbarrdahl replied:
I'm still learning solidity so this is more a question to understand it better :)

The code keeps track of guilds with the mapping `guildStructs` and an array of addresses (`_guilds`). When a guild is removed the array is cleaned up with `_guilds.pop` but the struct is still there. Could this be a problem as the number of guilds are created? Is it possible (or useful) to:
`delete guildStructs[guildAddress]`

rossneilson replied:
We could maybe consider just having it be a simple address  => uint256 mapping if we don't end up adding more attributes here which Im not certain we will

AugustoL replied:
What @carlbarrdahl is suggesting is to also remove the guild in the guildsStruct that is indexing it by index, you can free up some storage of the contract and it will be a bit cheaper to add information later, you are removing it in the array but not in the index.

AugustoL commented on contracts/erc20guild/utils/GuildRegistry.sol:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    address[] private _guilds;
+    Counters.Counter private _index;
```
no need to use private variables, there is no private information here and no need to protect this variables, use public instead.

AugustoL commented on contracts/erc20guild/utils/GuildRegistry.sol:
```diff
@@ -0,0 +1,48 @@
+    address[] private _guilds;
+    Counters.Counter private _index;
+
+    mapping(address => Guild) guildStructs;
```
The Guild is a struct, but this is an index, I would recommend naming it `guildsByAddress`, so when you read teh variable name you know that you can access the guild information in the mapping by just the address.

AugustoL commented on contracts/erc20guild/utils/GuildRegistry.sol:
```diff
@@ -0,0 +1,48 @@
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
```
Can you add natspect comments? the ERC20Guild has them, yo can also take a look here https://docs.soliditylang.org/en/v0.5.8/natspec-format.html

AugustoL commented on test/erc20guild/utils/GuildRegistry.js:
```diff
@@ -0,0 +1,105 @@
+    guildRegistry = await GuildRegistry.new({
+      from: accounts[0],
+    });
+    return {
```
unnecessary return here ?

AugustoL reviewed (changes requested):
I requested some changes, but the code looks good, great job :)

After this PR it would be good to add other test where a guild is added and remove with dxvote proposals :)

Kenny-Gin1 commented:
> I requested some changes, but the code looks good, great job :)
> 
> After this PR it would be good to add other test where a guild is added and remove with dxvote proposals :)

Thank you Augusto, appreciate it 👍. I will look at the comments and see if I can make the changes!

Kenny-Gin1 commented:
The latest commit includes:

- Removed the `Guild struct` since we are not sure if we want to add more attributes to the `GuildRegistry`. @rossneilson 
- Changed variable names and made certain variables public
- Added natspec comment as per @AugustoL suggestion above
